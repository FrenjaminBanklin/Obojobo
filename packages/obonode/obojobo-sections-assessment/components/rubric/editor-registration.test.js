jest.mock('obojobo-document-engine/src/scripts/oboeditor/util/keydown-util')
jest.mock('slate-react')
jest.mock('./editor-component', () => global.mockReactComponent(this, 'Rubric'))
jest.mock('./mod', () => global.mockReactComponent(this, 'Mod'))
jest.mock('./mod-list', () => global.mockReactComponent(this, 'ModList'))
jest.mock('./schema', () => ({ mock: 'schema' }))
jest.mock('./converter', () => ({ mock: 'converter' }))

import KeyDownUtil from 'obojobo-document-engine/src/scripts/oboeditor/util/keydown-util'
import SlateReact from 'slate-react'
import Rubric from './editor-registration'

const RUBRIC_NODE = 'ObojoboDraft.Sections.Assessment.Rubric'
const MOD_NODE = 'ObojoboDraft.Sections.Assessment.Rubric.Mod'
const MOD_LIST_NODE = 'ObojoboDraft.Sections.Assessment.Rubric.ModList'

describe('Rubric editor', () => {
	test('plugins.onPaste deals with no rubric', () => {
		const editor = {
			value: {
				blocks: [
					{
						key: 'mockBlockKey'
					}
				],
				document: {
					getClosest: () => false
				}
			}
		}
		editor.insertBlock = jest.fn().mockReturnValueOnce(editor)

		const event = {
			preventDefault: jest.fn()
		}

		Rubric.plugins.onPaste(event, editor, jest.fn())

		expect(event.preventDefault).not.toHaveBeenCalled()
	})

	test('plugins.onPaste deals with pasting into rubric', () => {
		const editor = {
			value: {
				blocks: [
					{
						key: 'mockBlockKey'
					}
				],
				document: {
					getClosest: (key, funct) => {
						funct({ type: 'mockType' })
						return true
					}
				}
			}
		}
		editor.insertText = jest.fn().mockReturnValueOnce(editor)

		SlateReact.getEventTransfer.mockReturnValueOnce({ text: 'mock text' })

		const event = {
			preventDefault: jest.fn()
		}

		Rubric.plugins.onPaste(event, editor, jest.fn())

		expect(editor.insertText).toHaveBeenCalled()
	})

	test('plugins.onCut deals with no rubric', () => {
		const editor = {
			value: {
				blocks: [
					{
						key: 'mockBlockKey'
					}
				],
				document: {
					getClosest: () => false
				}
			},
			extractTextToFragment: jest.fn()
		}
		editor.insertBlock = jest.fn().mockReturnValueOnce(editor)

		const event = {
			preventDefault: jest.fn()
		}

		Rubric.plugins.onCut(event, editor, jest.fn())

		expect(editor.extractTextToFragment).not.toHaveBeenCalled()
		expect(KeyDownUtil.deleteNodeContents).not.toHaveBeenCalled()
	})

	test('plugins.onCut deals with cutting from rubric', () => {
		const editor = {
			value: {
				blocks: [
					{
						key: 'mockBlockKey'
					}
				],
				document: {
					getClosest: (key, funct) => {
						funct({ type: 'mockType' })
						return true
					}
				},
				fragment: { text: 'selected text' }
			},
			extractTextToFragment: jest.fn()
		}

		const event = {
			preventDefault: jest.fn()
		}

		Rubric.plugins.onCut(event, editor, jest.fn())

		expect(editor.extractTextToFragment).toHaveBeenCalled()
		expect(KeyDownUtil.deleteNodeContents).toHaveBeenCalled()
	})

	test('plugins.onCopy deals with no rubric', () => {
		const editor = {
			value: {
				blocks: [
					{
						key: 'mockBlockKey'
					}
				],
				document: {
					getClosest: () => false
				}
			},
			extractTextToFragment: jest.fn()
		}

		const event = {
			preventDefault: jest.fn()
		}

		Rubric.plugins.onCopy(event, editor, jest.fn())

		expect(editor.extractTextToFragment).not.toHaveBeenCalled()
	})

	test('plugins.onCopy deals with copying from rubric', () => {
		const editor = {
			value: {
				blocks: [
					{
						key: 'mockBlockKey'
					}
				],
				document: {
					getClosest: (key, funct) => {
						funct({ type: 'mockType' })
						return true
					}
				},
				fragment: { text: 'selected text' }
			},
			extractTextToFragment: jest.fn()
		}

		const event = {
			preventDefault: jest.fn()
		}

		Rubric.plugins.onCopy(event, editor, jest.fn())

		expect(editor.extractTextToFragment).toHaveBeenCalled()
	})

	test('plugins.renderNode renders the rubric when passed', () => {
		const props = {
			attributes: { dummy: 'dummyData' },
			node: {
				type: RUBRIC_NODE,
				data: {
					get: () => {
						return {}
					}
				}
			}
		}

		expect(Rubric.plugins.renderNode(props, null, jest.fn())).toMatchSnapshot()
	})

	test('plugins.renderNode renders a modlist when passed', () => {
		const props = {
			attributes: { dummy: 'dummyData' },
			node: {
				type: MOD_LIST_NODE,
				data: {
					get: () => {
						return {}
					}
				}
			}
		}

		expect(Rubric.plugins.renderNode(props, null, jest.fn())).toMatchSnapshot()
	})

	test('plugins.renderNode renders a modlist when passed', () => {
		const props = {
			attributes: { dummy: 'dummyData' },
			node: {
				type: MOD_LIST_NODE,
				data: {
					get: () => {
						return {}
					}
				}
			}
		}

		expect(Rubric.plugins.renderNode(props, null, jest.fn())).toMatchSnapshot()
	})

	test('plugins.renderNode renders a modlist when passed', () => {
		const props = {
			attributes: { dummy: 'dummyData' },
			node: {
				type: MOD_NODE,
				data: {
					get: () => {
						return {}
					}
				}
			}
		}

		expect(Rubric.plugins.renderNode(props, null, jest.fn())).toMatchSnapshot()
	})

	test('plugins.renderNode calls next', () => {
		const props = {
			attributes: { dummy: 'dummyData' },
			node: {
				type: 'mockNode',
				data: {
					get: () => {
						return {}
					}
				}
			}
		}

		const next = jest.fn()

		expect(Rubric.plugins.renderNode(props, null, next)).toMatchSnapshot()
		expect(next).toHaveBeenCalled()
	})

	test('extractTextToFragment copies the current text', () => {
		const editor = {
			value: {
				fragment: { text: 'selected text' }
			}
		}
		expect(Rubric.plugins.queries.extractTextToFragment(editor)).toMatchSnapshot()
	})
})
