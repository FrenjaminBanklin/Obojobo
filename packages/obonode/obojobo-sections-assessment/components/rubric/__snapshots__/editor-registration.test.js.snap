// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rubric editor extractTextToFragment copies the current text 1`] = `
Immutable.Record {
  "data": Immutable.Map {},
  "key": "0",
  "nodes": Immutable.List [
    Immutable.Record {
      "data": Immutable.Map {},
      "key": "1",
      "nodes": Immutable.List [
        Immutable.Record {
          "data": Immutable.Map {
            "indent": 0,
          },
          "key": "2",
          "nodes": Immutable.List [
            Immutable.Record {
              "leaves": Immutable.List [
                Immutable.Record {
                  "marks": Immutable.Set [],
                  "text": "selected text",
                },
              ],
              "key": "3",
            },
          ],
          "type": "ObojoboDraft.Chunks.Text.TextLine",
        },
      ],
      "type": "ObojoboDraft.Chunks.Text",
    },
  ],
}
`;

exports[`Rubric editor plugins.renderNode calls next 1`] = `undefined`;

exports[`Rubric editor plugins.renderNode renders a modlist when passed 1`] = `
<ModList
  attributes={
    Object {
      "dummy": "dummyData",
    }
  }
  dummy="dummyData"
  node={
    Object {
      "data": Object {
        "get": [Function],
      },
      "type": "ObojoboDraft.Sections.Assessment.Rubric.ModList",
    }
  }
/>
`;

exports[`Rubric editor plugins.renderNode renders a modlist when passed 2`] = `
<ModList
  attributes={
    Object {
      "dummy": "dummyData",
    }
  }
  dummy="dummyData"
  node={
    Object {
      "data": Object {
        "get": [Function],
      },
      "type": "ObojoboDraft.Sections.Assessment.Rubric.ModList",
    }
  }
/>
`;

exports[`Rubric editor plugins.renderNode renders a modlist when passed 3`] = `
<Mod
  attributes={
    Object {
      "dummy": "dummyData",
    }
  }
  dummy="dummyData"
  node={
    Object {
      "data": Object {
        "get": [Function],
      },
      "type": "ObojoboDraft.Sections.Assessment.Rubric.Mod",
    }
  }
/>
`;

exports[`Rubric editor plugins.renderNode renders the rubric when passed 1`] = `
<Rubric
  attributes={
    Object {
      "dummy": "dummyData",
    }
  }
  dummy="dummyData"
  node={
    Object {
      "data": Object {
        "get": [Function],
      },
      "type": "ObojoboDraft.Sections.Assessment.Rubric",
    }
  }
/>
`;
